File: packages/sdk/src/migrations/migrateSnapshot.ts
import { Snapshot } from '@river-build/proto'
import { snapshotMigration0000 } from './snapshotMigration0000'
import { snapshotMigration0001 } from './snapshotMigration0001'

const SNAPSHOT_MIGRATIONS = [snapshotMigration0000, snapshotMigration0001]

export function migrateSnapshot(snapshot: Snapshot): Snapshot {
    const currentVersion = SNAPSHOT_MIGRATIONS.length
    if (snapshot.snapshotVersion >= currentVersion) {
        return snapshot
    }
    let result = snapshot
    for (let i: number = snapshot.snapshotVersion; i < currentVersion; i++) {
        result = SNAPSHOT_MIGRATIONS[i](result)
    }
    result.snapshotVersion = currentVersion
    return result
}

File: core/node/events/migrations/migrate_snapshot.go
package migrations

import (
	. "github.com/river-build/river/core/node/protocol"
)

type migrationFunc func(*Snapshot) *Snapshot

// should be kept in sync with packages/sdk/src/migrations/migrate_snapshot.ts
var MIGRATIONS = []migrationFunc{
	snapshot_migration_0000,
	snapshot_migration_0001,
}

func CurrentSnapshotVersion() int32 {
	return int32(len(MIGRATIONS))
}

func MigrateSnapshot(iSnapshot *Snapshot) *Snapshot {
	currentVersion := CurrentSnapshotVersion()
	if iSnapshot.SnapshotVersion >= currentVersion {
		return iSnapshot
	}
	for i := iSnapshot.SnapshotVersion; i < currentVersion; i++ {
		iSnapshot = MIGRATIONS[i](iSnapshot)
	}
	iSnapshot.SnapshotVersion = currentVersion
	return iSnapshot
}
